testsuite/testcases/src/validator_join_leave_test.rs

```diff
@@ -0,0 +1,223 @@
+            )
+            .await
+        });
+
+        let mut public_info = swarm.chain_info().into_aptos_public_info();
+
+        let mut validator_cli_indices = Vec::new();
+
+        let starting_seed_in_decimal = i64::from_str_radix(FORGE_KEY_SEED, 16)?;
+
+        for i in 0..num_validators {
+            // Initialize keyGen to get validator private keys. We uses the same seed in the test
+            // driver as in the genesis script so that the validator keys are deterministic.
+            let mut seed_slice = [0u8; 32];
+            let seed_in_decimal = starting_seed_in_decimal + (i as i64);
+            let seed_in_hex_string = format!("{seed_in_decimal:0>64x}");
+
+            hex::decode_to_slice(seed_in_hex_string, &mut seed_slice)?;
+
+            let mut keygen = KeyGen::from_seed(seed_slice);
+
+            let (validator_cli_index, _keys, account_balance) = runtime.block_on(async {
+                let (validator_cli_index, keys) =
+                    init_validator_account(&mut cli, &mut keygen).await;
+
+                let auth_key = AuthenticationKey::ed25519(&keys.account_private_key.public_key());
+                let validator_account_address = AccountAddress::new(*auth_key.derived_address());
+
+                public_info
+                    .mint(validator_account_address, DEFAULT_FUNDED_COINS)
+                    .await
+                    .unwrap();
+
+                let account_balance = public_info
+                    .get_balance(validator_account_address)
+                    .await
+                    .unwrap();
+
+                (validator_cli_index, keys, account_balance)
+            });
+            assert_eq!(account_balance, DEFAULT_FUNDED_COINS);
+            validator_cli_indices.push(validator_cli_index);
+
+            assert_eq!(
+                runtime.block_on(get_validator_state(&cli, validator_cli_index)),
+                ValidatorState::ACTIVE
+            );
+        }
+
+        // Log the test setup
+        info!(
+            "Running validator join and leave test {:?} with {:?} validators.",
+            self.name(),
+            num_validators,
+        );
+
+        // Wait for all nodes to synchronize and stabilize.
+        info!("Waiting for the validators to be synchronized.");
+        runtime.block_on(async {
+            swarm
+                .wait_for_all_nodes_to_catchup(Duration::from_secs(MAX_NODE_LAG_SECS))
+                .await
+        })?;
+
+        // Wait for 1/3 of the test duration.
+        std::thread::sleep(duration / 3);
+
+        // 1/3 validators leave the validator set.
+        info!("Make the last 1/3 validators leave the validator set!");
+        for operator_index in validator_cli_indices.iter().rev().take(num_validators / 3) {
+            runtime.block_on(async {
+                cli.leave_validator_set(*operator_index, None)
+                    .await
+                    .unwrap();
+
+                reconfig(
```

Comment:
oh, you should probably do reconfig once at the end, this will create bunch of epochs.

we also want to test that they can all ask to leave simultaneously


in the follow-up randomized test you are mentioning below, you can randomly select how many validators to leave simulatenously, before reconfig.

Reply:
Added a reconfig after the loop.